### PR TITLE
BUG: Fix numpy boolean subtraction error in Series.diff

### LINKED ISSUE #17294: pd.Series.diff() on boolean values
#### Code Sample

```python
>> x = pd.Series([True,False,True])
>> x.diff()
0     NaN
1    True
2    True
dtype: object
>> x - x.shift()
0    NaN
1     -1
2      1
dtype: object
```
#### Problem description

It's counter-intuitive that the results of above are different.

The current implementation of pd.Series.diff uses algorithms.diff that subtracts 2 numpy arrays in the following way

```python
out_arr[res_indexer] = arr[res_indexer] - arr[lag_indexer]
```

As pointed [here](https://github.com/numpy/numpy/issues/9251) such behaviour is deprecated in favor to np.diff. But np.diff also treats booleans in binary operations in its own numpy way, that is different from native python (replace False with 0, replace True with 1).

```python
>> np.array([True, False]) - np.array([False, True])
/home/deoxys/miniconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: DeprecationWarning: numpy boolean subtract, the `-` operator, is deprecated, use the bitwise_xor, the `^` operator, or the logical_xor function instead.
array([ True,  True], dtype=bool)
>> np.diff([False, True, False])
array([ True,  True], dtype=bool)
>> True - False
1
>> False - True
-1
```

#### Expected Output

I believe there is no correct way of subtracting booleans. But, it's definitely strange that operations like `x - x.shift()` and `x.diff()` provide different results.

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------

commit: None
python: 3.6.0.final.0
python-bits: 64
OS: Linux
OS-release: 4.11.9-1-ARCH
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.20.1
pytest: None
pip: 9.0.1
setuptools: 27.2.0
Cython: 0.25.2
numpy: 1.12.1
scipy: 0.19.0
xarray: None
IPython: 6.0.0
sphinx: None
patsy: None
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.0.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: 4.6.0
html5lib: 0.999
sqlalchemy: None
pymysql: None
psycopg2: 2.7.1 (dt dec pq3 ext lo64)
jinja2: 2.9.6
s3fs: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by Unprocessable**
This fixes #27755, for more than three years now NumPy has not allowed the subtraction of boolean series. 

TypeError                                 Traceback (most recent call last)
<ipython-input-46-3da3b949c6bd> in <module>
      1 data = pd.Series([0,-1,-2,-3,-4,-3,-2,-1,0,-1,-1,0,-1,-2,-3,-2,0])
      2 filtered = data.between(-2,0, inclusive = True)
----> 3 filtered.diff()
      4 print(filtered)

~\AppData\Local\Continuum\anaconda3\lib\site-packages\pandas\core\series.py in diff(self, periods)
   2191         dtype: float64
   2192         """
-> 2193         result = algorithms.diff(com.values_from_object(self), periods)
   2194         return self._constructor(result, index=self.index).__finalize__(self)
   2195 

~\AppData\Local\Continuum\anaconda3\lib\site-packages\pandas\core\algorithms.py in diff(arr, n, axis)
   1817             out_arr[res_indexer] = result
   1818         else:
-> 1819             out_arr[res_indexer] = arr[res_indexer] - arr[lag_indexer]
   1820 
   1821     if is_timedelta:

TypeError: numpy boolean subtract, the `-` operator, is deprecated, use the bitwise_xor, the `^` operator, or the logical_xor function instead.

- [x] closes #17294
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
tests are always the first thing we write, pls add them

File: pandas/tests/series/test_diff.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+from pandas import (
+    Series
+)
+
+from numpy import nan
+
+
+def test_diff():
+    data = Series(
```
Comment (TomAugspurger) [inline pandas/tests/series/test_diff.py]: Does it need to be this long?

File: pandas/tests/series/test_diff.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        [0, -1, -2, -3, -4, -3, -2, -1, 0, -1, -1, 0, -1, -2, -3, -2, 0]
+    )
+    filtered = data.between(-2, 0, inclusive=True)
+    diff_boolean = filtered.diff()
```
Comment (TomAugspurger) [inline pandas/tests/series/test_diff.py]: Can you remove this and just make the assert on the the result of `data.between`? I'm not sure why `diff` is needed.

Comment (TomAugspurger) [reply pandas/tests/series/test_diff.py]: And use `tm.assert_series_equal` instead of `assert`

**Review by TomAugspurger — COMMENTED** [review]
Needs a release note. What version of NumPy will / does this show up in?

File: pandas/tests/series/test_diff.py (hunk 2 above)
Comment (Unprocessable) [reply pandas/tests/series/test_diff.py]: Alright, first comment, release notes, do you have a link for how to write that?
Second comment, agreed, I'll shorten it.
Third comment, the diff is what I used to test the feature, it causes the issue in my test build (diff on boolean series). So if I remove it, the bug will not show up when testing.
Last comment, I tried using pytest and got some errors so went with the old fashioned way.

Comment (Unprocessable) [reply pandas/tests/series/test_diff.py]: Apparently the warning was added in numpy 1.13.0, 3 years ago.

Comment (TomAugspurger) [reply pandas/tests/series/test_diff.py]: So is the issue with `Series.between` or `Series.diff`? Let's use one or the other to keep the test focused.

**Comment by jbrockmendel** [comment]
Are there other tests for diff?  If so they should be collected here or this should join them.

File: pandas/tests/series/test_diff.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
+from numpy import nan
+from pandas import Series, date_range
+from pandas.tests.series.common import TestData
+from pandas.core.indexes.timedeltas import TimedeltaIndex
+from pandas.util.testing import assert_series_equal
+
+
+class TestDiff(TestData):
```
Comment (jreback) [inline pandas/tests/series/test_diff.py]: you can drop the class TestDiff
use a fixture for ts here (it’s likely already a fixture)

can you split this up into separate tests / paramaterize

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/series/test_diff.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        # with tz
+        s = Series(
+            date_range("2000-01-01 09:00:00", periods=5, tz="US/Eastern"), name="foo"
+        )
+        result = s.diff()
+        assert_series_equal(
+            result, Series(TimedeltaIndex(["NaT"] + ["1 days"] * 4), name="foo")
+        )
+        
+        # boolean series
```
Comment (TomAugspurger) [inline pandas/tests/series/test_diff.py]: Can you add a test for a mix of bool & NA? `pd.Series([True, False, np.nan, False], dtype="object')`

File: pandas/tests/series/test_diff.py (hunk 2 above)
Comment (Unprocessable) [reply pandas/tests/series/test_diff.py]: Fixed

File: pandas/tests/series/test_diff.py (hunk 4 above)
Comment (Unprocessable) [reply pandas/tests/series/test_diff.py]: Done

File: pandas/tests/series/test_diff.py (hunk 3 above)
Comment (Unprocessable) [reply pandas/tests/series/test_diff.py]: I would love to, but I will learn that after the merge, I do not have the time now to fix it. Rewriting the whole file will take a while, let alone that the results of the test should always be the same, which cannot be guaranteed after a rewrite.

**Comment by Unprocessable** [comment]
I cannot get the import order fixed correctly, on my PC isort does not show any changes, but the checks do show up.

**Comment by TomAugspurger** [comment]
For now I think it's fine to just add the new test to pandas.tests.series.test_analytics.TestSeriesAnalytics. There's a `test_npdiff` there that's related.

**Comment by TomAugspurger** [comment]
@Unprocessable do you have time to update with the recommendations from https://github.com/pandas-dev/pandas/pull/27755#issuecomment-519275906?

**Comment by Unprocessable** [comment]
Apparently there are some problems with tslibs. Apart from that it should be fine. @jbrockmendel @TomAugspurger Feel free to comment on it again or merge it. I might not get around to fixing it in the next few weeks and will definitely not re-factor a whole code base to use functions/fixtures. For a quick fix however, I'll try to make the time.

**Comment by Unprocessable** [comment]
Just checked it, everyone seems to have that problem with pandas-master at the moment, so it is not due to my code. Please fix the master and rerun the tests if you're unsure, otherwise feel free to merge. Let's finally fix the three year old bug!

**Comment by TomAugspurger** [comment]
Hopefully fixed in https://github.com/pandas-dev/pandas/pull/27845.

If you can merge master & re-push it should be fixed.

**Comment by Unprocessable** [comment]
> 
> Hopefully fixed in #27845. If you can merge master & re-push it should be fixed.
> 
> I might have some time tomorrow for some quick fixes if required. 
No, I hate to say this, but I guess you are in a different time zone and tomorrow means something different. Sorry, this will have to wait a few weeks or it will have to go as is.

**Comment by TomAugspurger** [comment]
No rush, thanks.

**Comment by simonjayhawkins** [comment]
superseded by #28251

**Pull request closed without merging**